packages/gatsby/src/utils/api-runner-node.js

```diff
@@ -527,9 +527,25 @@ module.exports = async (api, args = {}, { pluginSource, activity } = {}) =>
         return null
       }
 
+      let gatsbyNode = pluginNodeCache.get(plugin.name)
+      if (!gatsbyNode) {
+        gatsbyNode = require(`${plugin.resolve}/gatsby-node`)
+        pluginNodeCache.set(plugin.name, gatsbyNode)
+      }
+
       const pluginName =
         plugin.name === `default-site-plugin` ? `gatsby-node.js` : plugin.name
 
+      // TODO: rethink createNode API to handle this better
+      if (
+        api === `onCreateNode` &&
+        gatsbyNode?.shouldOnCreateNode && // Don't bail if this api is not exported
+        !gatsbyNode.shouldOnCreateNode(args.node, plugin.pluginOptions)
```

Comment:
Not a blocker - just to not miss this part:

skipping `runAPI` makes this not behave like rest of APIs (which is also kind of the point of it), but it changes few things (like right now passing `node` plainly as first arg as opposed to passing object with `node` like actual `onCreateNode` and also not passing rest of "shared" things - https://www.gatsbyjs.com/docs/node-api-helpers/).

I don't think we need all of those helpers in the "test" callback like that (it should be quick check and not rely on anything else other than `node` itself), but it does introduce first "odd" `gatsby-node` hook and so we also need to think about user expectations (and our docs)

Reply:
@pieh @wardpeet blocking or not? :)

Reply:
🤔 I agree with @pieh so let's do it differently. Let's call it `unstable_shouldOnCreateNode({ node: args.node }, plugin.pluginOptions)` so we're unblocked and it shows that it's not a complete API.

Reply:
I like that. Will roll with that

Reply:
eslint does not approve. Will go with `unstableShouldOnCreateNode` instead, otherwise every mention of it will also require an eslint suppression line.

Reply:
plot twist: eslint has been taught a lesson instead